mate-conf/gsettings/gsettings-data-convert.1
.\"     Title: gsettings-data-convert
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: 05/05/2010
.\"    Manual: User Commands
.\"    Source: User Commands
.\"
.TH "GSETTINGS\-DATA\-CON" "1" "05/05/2010" "User Commands" "User Commands"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
gsettings-data-convert \- MateConf to GSettings data migration
.SH "SYNOPSIS"
.HP 23
\fBgsettings\-data\-convert\fR [option...]
.SH "DESCRIPTION"
.PP

\fBgsettings\-data\-convert\fR
reads values out of the users MateConf database and stores them in GSettings\&.
.PP
The information about the mapping from MateConf keys to GSettings keys is taken from files in
\fI/usr/share/MateConf/gsettings\fR\&. Each file in that directory is read as a key file, with sections for each GSettings schema that is being converted\&. The entries in each section map GSettings keys to paths in the MateConf database\&. Currently,
\fBgsettings\-data\-convert\fR
supports all basic MateConf types as well as lists of strings and lists of integers\&.
.PP
A simple example might look like this:
.sp
.RS 4
.nf
[org\&.mate\&.fonts]
antialiasing = /desktop/mate/font_rendering/antialiasing
dpi = /desktop/mate/font_rendering/dpi
hinting = /desktop/mate/font_rendering/hinting
rgba_order = /desktop/mate/font_rendering/rgba_order
.fi
.RE
.PP
If the GSettings schema does not have a fixed path, you must specify the path in the group name, separated by a colon character:
.sp
.RS 4
.nf
[org\&.mate\&.fonts:/desktop/mate/font\-rendering/]
.fi
.RE
.PP

\fBgsettings\-data\-convert\fR
keeps a list of the key files it has already converted, so it is safe to run it repeatedly to handle newly appeared key files\&. The expected use of this utility is to make each application install a key file for the MateConf keys that it wants to be migrated, and run
\fBgsettings\-data\-convert\fR
every time a user logs in\&.
.SS "Options"
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Print help and exit
.RE
.PP
\fB\-\-dry\-run\fR
.RS 4
Do not make any changes, just report what would be done
.RE
.PP
\fB\-\-verbose\fR
.RS 4
Show verbose messages
.RE
.SH "SEE ALSO"
.PP

\fBgsettings-schema-convert\fR(1)
is a related command that helps with the conversion of schemas from MateConf to GSettings\&.


libmate/NEWS
libmate 2.32.0

mate-sound APIs now use libcanberra internally. That means that
mate_sound_connection_get() and will always return error (-1), and
mate_sound_sample_load also always return -1 but still cache
the sample using the libcanberra API.
libmate continues to link to libesound and libaudiofile for ABI
compatibility; use --disable-esd to remove this dependency.
(#603198, #603349)


libmate 2.30.0

* Enable silent build using AM_SILENT_RULES (Christian Persch)
* Put SlowKeys and BounceKeys default delays back to 300 (Willie Walker)
* Deprecate mate-sound (Christian Persch)
* Deprecate mate_[un]setenv (Christian Persch)
* Update the description of the picture_options key (Matthias Clasen)

* Updated translations
  - Khaled Hosny (ar)
  - astur (ast)
  - Alexander Shopov (bg)
  - Sadia Afroz (bn)
  - Jordi Mallach (ca)
  - Reşat SABIQ (crh)
  - Andre Klapper (cs)
  - Iestyn Pryce (cy)
  - Ask Hjorth Larsen (da)
  - Mario Blättermann (de)
  - Bakaoukas Nikolaos (el)
  - Thomas Thurman (en@shaw)
  - Bruce Cowan (en_GB)
  - Jorge González (es)
  - Mattias Põldaru (et)
  - Iñaki Larrañaga Murgoitio (eu)
  - Tommi Vainikainen (fi)
  - Claude Paroz (fr)
  - Fran Diéguez (gl)
  - Gabor Kelemen (hu)
  - Francesco Marletta (it)
  - Rimas Kudelis (lt)
  - Badral Sanligiin (mn)
  - Kjartan Maraas (nb)
  - Nils-Christoph Fiedler (nds)
  - Wouter Bolsterlee (nl)
  - A S Alam (pa)
  - Piotr Drąg (pl)
  - Duarte Loreto (pt)
  - Henrique P Machado (pt_BR)
  - Adi Roiban (ro)
  - Yuri Kozlov (ru)
  - Marcel Telka (sk)
  - Matej Urbančič (sl)
  - Горан Ракић (sr)
  - Goran Rakić (sr@latin)
  - Daniel Nylander (sv)
  - Maxim Dziumanenko (uk)
  - Clytie Siddall (vi)
  - Chao-Hsiung Liao (zh_HK)
  - Chao-Hsiung Liao (zh_TW)


libmate 2.28.0

Aron Xu (1):
      Corrected 5 strings.
Denis Arnaud (1):
      Updated breton translation and added br in LINGUAS
Lucian Adrian Grijincu (1):
      Updated Romanian translation
Miloš Popović (1):
      Updated Serbian translation
Rodrigo L. M. Flores (1):
      Updated Brazilian Portuguese mailing list address.

 
libmate 2.27.5

* Update default properties for menus_have_icons, buttons_have_icons and
  toolbar_style
* Port mate-sound.c to g_poll to fix portability for old OSX versions without
  poll.
* Update the address of the FSF by replacing old bits with new ones

* Updated translations
  - Jordi Mallach (ca)
  - Jordi Mallach (ca@valencia)
  - Khaled Hosny (ar)
  - Kjartan Maraas (nb)
  - Marcel Telka (sk)
  - Ray Wang (zh_CN)
  - Shankar Prasad (kn)
  - Sweta Kothari (gu)


libmate 2.26.0

* Updated translations
  (as, bn_IN, crh, cs, da, el, gl, hi, it, ja, lt, ml, mr,
   ro, ru, ta, te)

libmate 2.24.1

* Fix warnings about missing format specifiers (Kjartan)
* Remove #if 0'ed out code (Kjartan)
* Updated translations
  (ar, bg, en_GB, eu, fr, gu, lv, nl, or, pl, pt, th, tr,
   vi, zh_HK, zh_TW)

libmate 2.24.0.1

* Make thumbnail housekeeping less agressive by default
  (Michael Chudobiak)

libmate 2.24.0

* Updated translations
  (ar, da, gu, it, ro, ru, sr, ta, te)

libmate 2.23.92

* Port mate-open to GIO. (Jean-Yves Lefort)
* Build fix (Tor Lillqvist)

libmate 2.23.5

* Fixes for sound related issues (Jeff Steadfast)

libmate 2.23.4

* Updated translations.

libmate 2.23.3

* Code cleanups (Kjartan, Christian)
* Added schema to support the thumbnail cache cleaner (Michael J. Chudobiak)
* Plug leak (Stefan Kost)
* Add disable_application_handlers lockdown schema (William Jon McCann)
 
libmate 2.22.0

* Release 2.22.0
 
libmate 2.21.91

* Add schemas for tasks and calendar
* Use the GIO backend for the filechooser

libmate 2.21.90

* Set sound as enabled by default (Christian)
* Bypass a11y on win32 (Tor)
* Remove unused popt stuff (dmacks at netspace org) 
* Fix the toolbar icon size schema type (#401030)
* Win32 fixes
* SElinux fixes
* Warning fixes
* Translation updates

libmate 2.20.0

* Build fixes (Christian Persch)
* Set new background (Lucas Rocha, Vincent Untz)
* Fix module loading (Mariano Suárez-Alvarez, Christian Persch)
* Only depend on esd when esd is compiled in (Daniel Gryniewicz)
* Updated translations:
  - Albanian (Laurent Dhima)
  - Arabic (Khaled Hosny)
  - Assamese (Amitakhya Phukan)
  - Basque (Inaki Larranaga Murgoitio)
  - Bengali (Khandakar Mujahidul Islam)
  - Brazilian Portugese (Og Maciel)
  - Bulgarian (Alexander Shopov)
  - Canadian English (Adam Weinberger)
  - Catalan (Gil Forcada)
  - Danish (Kenneth Nielsen)
  - Estonian (Ivar Smolin)
  - Finnish (Ilkka Tuohela)
  - French (Stephane Raimbault)
  - Georgian (Vladimer Sichinava)
  - German (Hendrik Richter)
  - Gujarati (Ankit Patel)
  - Hungarian (Gabor Kelemen)
  - Italian (Francesco Marletta)
  - Japanese (Takeshi AIHANA)
  - Lithuanian (Žygimantas Beručka)
  - Macedonian (Jovan Naumovski)
  - Malayalam (Ani Peter)
  - Norwegian bokmål (Kjartan Maraas)
  - Polish (MATE Pl Team)
  - Portugese (Duarte Loreto)
  - Romaninan (Mişu Moldovan)
  - Russian (Nikolay V. Shmyrev)
  - Serbian (Goran Rakić)
  - Sinhala (Danishka Navin)
  - Spanish (Jorge Gonzalez)
  - Swedish (Daniel Nylander)
  - Thai (Theppitak Karoonboonyanan)
  - Ukrainian (Maxim Dziumanenko)
  - Vietnamese (Clytie Siddall)

libmate 2.19.1

* Don't initialise ALSA when sound is disabled (449249; Christian Persch)
* Translation updates: Runa Bhattacharjee (bn_IN), Hendrik Richter (de),
  Iñaki Larrañaga Murgoitio (eu), Ilkka Tuohela (fi), Ankit Patel (gu),
  Takeshi AIHANA (ja), Leonardo Fereira Fontenelle (pt_BR), Clytie Siddall (vi)

libmate 2.19.0

What's new since 2.18.0

* Rebuild the type builtins when the Makefile changes (Christian Persch)
* Make bug-buddy work with applets again (Matthias Clasen, #424949)
* Fix problems when building static libraries (Kjartan Maraas, #405352)
* Fix typos (Kenneth Nielsen, Kjartan Maraas, #429751)
* Add more information to docs for mate_program_init () (Christian Persch,
  #82943)
* Updated translations
  - mg: Thierry Randrianiriana
  - nb: Kjartan Maraas

mate-conf/gsettings/gsettings-data-convert.c
/*
 * Copyright (C) 2010 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 * Author: Matthias Clasen <mclasen@redhat.com>
 */

#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>

#include <glib.h>
#include <gio/gio.h>
#include <mateconf/mateconf-client.h>

static const gchar convert_dir[] = DATADIR "/MateConf/gsettings";

static gboolean verbose = FALSE;
static gboolean dry_run = FALSE;

extern const gchar *mateconf_value_type_to_string (int type);

static gboolean
handle_file (const gchar *filename)
{
  GKeyFile *keyfile;
  MateConfClient *client;
  MateConfValue *value;
  gint i, j;
  gchar *mateconf_key;
  gchar **groups;
  gchar **keys;
  GVariantBuilder *builder;
  GVariant *v;
  const gchar *s;
  gchar *str;
  gint ii;
  GSList *list, *l;
  GSettings *settings;
  GError *error;

  keyfile = g_key_file_new ();

  error = NULL;
  if (!g_key_file_load_from_file (keyfile, filename, 0, &error))
    {
      g_printerr ("%s\n", error->message);
      g_error_free (error);

      g_key_file_free (keyfile);

      return FALSE;
    }

  client = mateconf_client_get_default ();

  groups = g_key_file_get_groups (keyfile, NULL);
  for (i = 0; groups[i]; i++)
    {
      gchar **schema_path;

      schema_path = g_strsplit (groups[i], ":", 2);

      if (verbose)
        {
          g_print ("collecting settings for schema '%s'\n", schema_path[0]);
          if (schema_path[1])
            g_print ("for storage at '%s'\n", schema_path[1]);
        }

      if (schema_path[1] != NULL)
        settings = g_settings_new_with_path (schema_path[0], schema_path[1]);
      else
        settings = g_settings_new (schema_path[0]);

      g_settings_delay (settings);

      error = NULL;
      if ((keys = g_key_file_get_keys (keyfile, groups[i], NULL, &error)) == NULL)
        {
          g_printerr ("%s", error->message);
          g_error_free (error);

          continue;
        }

      for (j = 0; keys[j]; j++)
        {
          if (strchr (keys[j], '/') != 0)
            {
              g_printerr ("Key '%s' contains a '/'\n", keys[j]);

              continue;
            }

          error = NULL;
          if ((mateconf_key = g_key_file_get_string (keyfile, groups[i], keys[j], &error)) ==  NULL)
            {
              g_printerr ("%s", error->message);
              g_error_free (error);

              continue;
            }

          error = NULL;
          if ((value = mateconf_client_get_without_default (client, mateconf_key, &error)) == NULL)
            {
              if (error)
                {
                  g_printerr ("Failed to get MateConf key '%s': %s\n",
                              mateconf_key, error->message);
                  g_error_free (error);
                }
              else
                {
                  if (verbose)
                    g_print ("Skipping MateConf key '%s', no user value\n",
                             mateconf_key);
                }

              g_free (mateconf_key);

              continue;
            }

          switch (value->type)
            {
            case MATECONF_VALUE_STRING:
              if (dry_run)
                g_print ("set key '%s' to string '%s'\n", keys[j],
                         mateconf_value_get_string (value));
              else
                g_settings_set (settings, keys[j], "s",
                                mateconf_value_get_string (value));
              break;

            case MATECONF_VALUE_INT:
              if (dry_run)
                g_print ("set key '%s' to integer '%d'\n",
                         keys[j], mateconf_value_get_int (value));
              else
                g_settings_set (settings, keys[j], "i",
                                mateconf_value_get_int (value));
              break;

            case MATECONF_VALUE_BOOL:
              if (dry_run)
                g_print ("set key '%s' to boolean '%d'\n",
                         keys[j], mateconf_value_get_bool (value));
              else
                g_settings_set (settings, keys[j], "b",
                                mateconf_value_get_bool (value));
              break;

            case MATECONF_VALUE_FLOAT:
              if (dry_run)
                g_print ("set key '%s' to double '%g'\n",
                         keys[j], mateconf_value_get_float (value));
              else
                g_settings_set (settings, keys[j], "d",
                                mateconf_value_get_float (value));
              break;

            case MATECONF_VALUE_LIST:
              switch (mateconf_value_get_list_type (value))
                {
                case MATECONF_VALUE_STRING:
                  builder = g_variant_builder_new (G_VARIANT_TYPE_ARRAY);
                  list = mateconf_value_get_list (value);
                  if (list != NULL)
                    {
                      for (l = list; l; l = l->next)
                        {
                          MateConfValue *lv = l->data;
                          s = mateconf_value_get_string (lv);
                          g_variant_builder_add (builder, "s", s);
                        }
                      v = g_variant_new ("as", builder);
                    }
                  else
                    v = g_variant_new_array (G_VARIANT_TYPE_STRING, NULL, 0);
                  g_variant_ref_sink (v);

                  if (dry_run)
                    {
                      str = g_variant_print (v, FALSE);
                      g_print ("set key '%s' to a list of strings: %s\n",
                               keys[j], str);
                      g_free (str);
                    }
                  else
                    g_settings_set_value (settings, keys[j], v);

                  g_variant_unref (v);
                  g_variant_builder_unref (builder);
                  break;

                case MATECONF_VALUE_INT:
                  builder = g_variant_builder_new (G_VARIANT_TYPE_ARRAY);
                  list = mateconf_value_get_list (value);
                  if (list != NULL)
                    {
                      for (l = list; l; l = l->next)
                        {
                          MateConfValue *lv = l->data;
                          ii = mateconf_value_get_int (lv);
                          g_variant_builder_add (builder, "i", ii);
                        }
                      v = g_variant_new ("ai", builder);
                    }
                  else
                    v = g_variant_new_array (G_VARIANT_TYPE_INT32, NULL, 0);
                  g_variant_ref_sink (v);

                  if (dry_run)
                    {
                      str = g_variant_print (v, FALSE);
                      g_print ("set key '%s' to a list of integers: %s\n",
                               keys[j], str);
                      g_free (str);
                    }
                  else
                    g_settings_set_value (settings, keys[j], v);

                  g_variant_unref (v);
                  g_variant_builder_unref (builder);
                  break;

                default:
                  g_printerr ("Keys of type 'list of %s' not handled yet\n",
                              mateconf_value_type_to_string (mateconf_value_get_list_type (value)));
                  break;
                }
              break;

            default:
              g_printerr ("Keys of type %s not handled yet\n",
                          mateconf_value_type_to_string (value->type));
              break;
            }

          mateconf_value_free (value);
          g_free (mateconf_key);
        }

      g_strfreev (keys);

      if (!dry_run)
        g_settings_apply (settings);

      g_object_unref (settings);
      g_strfreev (schema_path);
    }

  g_strfreev (groups);

  g_object_unref (client);

  return TRUE;
}

static void
load_state (time_t  *mtime,
            gchar ***converted)
{
  gchar *filename;
  GKeyFile *keyfile;
  GError *error;
  gchar *str;
  gchar **list;

  *mtime = 0;
  *converted = g_new0 (gchar *, 1);

  filename = g_build_filename (g_get_user_data_dir (), "gsettings-data-convert", NULL);
  keyfile = g_key_file_new ();

  /* ensure file exists */
  if (!g_file_test (filename, G_FILE_TEST_EXISTS))
    return;

  error = NULL;
  if (!g_key_file_load_from_file (keyfile, filename, 0, &error))
    {
      g_printerr ("%s\n", error->message);
      g_error_free (error);
      return;
    }

  error = NULL;
  if ((str = g_key_file_get_string (keyfile, "State", "timestamp", &error)) == NULL)
    {
      g_printerr ("%s\n", error->message);
      g_error_free (error);
    }
  else
    {
      *mtime = (time_t)g_ascii_strtoll (str, NULL, 0);
      g_free (str);
    }

  error = NULL;
  if ((list = g_key_file_get_string_list (keyfile, "State", "converted", NULL, &error)) == NULL)
    {
      g_printerr ("%s\n", error->message);
      g_error_free (error);
    }
  else
    {
      g_strfreev (*converted);
      *converted = list;
    }

  g_key_file_free (keyfile);
  g_free (filename);
}

static gboolean
save_state (time_t   mtime,
            gchar  **converted)
{
  gchar *filename;
  GKeyFile *keyfile;
  gchar *str;
  GError *error;
  gboolean result;

  /* Make sure the state directory exists */
  if (g_mkdir_with_parents (g_get_user_data_dir (), 0755))
    {
      g_printerr ("Failed to create directory %s: %s\n",
                  g_get_user_data_dir (), g_strerror (errno));
      return FALSE;
    }

  filename = g_build_filename (g_get_user_data_dir (), "gsettings-data-convert", NULL);
  keyfile = g_key_file_new ();

  str = g_strdup_printf ("%ld", mtime);
  g_key_file_set_string (keyfile,
                         "State", "timestamp", str);
  g_free (str);

  g_key_file_set_string_list (keyfile,
                              "State", "converted",
                              (const gchar * const *)converted, g_strv_length (converted));

  str = g_key_file_to_data (keyfile, NULL, NULL);
  g_key_file_free (keyfile);

  error = NULL;
  if (!g_file_set_contents (filename, str, -1, &error))
    {
      g_printerr ("%s\n", error->message);
      g_error_free (error);

      result = FALSE;
    }
  else
    result = TRUE;

  g_free (filename);
  g_free (str);

  return result;
}

int
main (int argc, char *argv[])
{
  time_t stored_mtime;
  time_t dir_mtime;
  struct stat statbuf;
  GError *error;
  gchar **converted;
  GDir *dir;
  const gchar *name;
  gchar *filename;
  gint i;
  GOptionContext *context;
  GOptionEntry entries[] = {
    { "verbose", 0, 0, G_OPTION_ARG_NONE, &verbose, "show verbose messages", NULL },
    { "dry-run", 0, 0, G_OPTION_ARG_NONE, &dry_run, "do not perform any changes", NULL },
    { NULL }
  };

  context = g_option_context_new ("");

  g_option_context_set_summary (context,
    "Migrate settings from the users MateConf database to GSettings.");

  g_option_context_add_main_entries (context, entries, NULL);

  error = NULL;
  if (!g_option_context_parse (context, &argc, &argv, &error))
    {
      g_printerr ("%s\n", error->message);
      return 1;
    }

  load_state (&stored_mtime, &converted);

  /* If the directory is not newer, exit */
  if (stat (convert_dir, &statbuf) == 0)
    dir_mtime = statbuf.st_mtime;
 else
    {
      if (verbose)
        g_print ("Directory '%s' does not exist, nothing to do\n", convert_dir);
      return 0;
    }

  if (dir_mtime <= stored_mtime)
    {
      if (verbose)
        g_print ("All uptodate, nothing to do\n");
      return 0;
    }

  error = NULL;
  dir = g_dir_open (convert_dir, 0, &error);
  if (dir == NULL)
    {
      g_printerr ("Failed to open '%s': %s\n", convert_dir, error->message);
      return 1;
    }

  while ((name = g_dir_read_name (dir)) != NULL)
    {
       for (i = 0; converted[i]; i++)
         {
           if (strcmp (name, converted[i]) == 0)
             {
               if (verbose)
                 g_print ("File '%s already converted, skipping\n", name);
               goto next;
             }
         }

      filename = g_build_filename (convert_dir, name, NULL);

      if (handle_file (filename))
        {
          gint len;

          /* Add the the file to the converted list */
          len = g_strv_length (converted);
          converted = g_realloc (converted, (len + 2) * sizeof (gchar *));
          converted[len] = g_strdup (name);
          converted[len + 1] = NULL;
        }

      g_free (filename);

 next: ;
    }

  if (!dry_run)
    {
      if (!save_state (dir_mtime, converted))
        return 1;
    }

  return 0;
}

